Add documentation on Perl packages. (#1790)
conda-forge/conda-forge.github.io: src/maintainer/knowledge_base.rst

So make sure that you **tag** your package in such a way that the package name
that conda-build spits out will sort the package uploaded with an ``rc`` label
higher than the package uploaded with the ``dev`` label.

Perl packages
=============

Perl has three standard install locations: **core**, **vendor**, and **site**. Here, **core** is used only for the perl itself, **vendor** is used for installing any other conda-forge packages, and **site** is used for the user to install things locally from sources other than conda-forge.

> The most commonly used build system for `perl packages <https://github.com/conda-forge/perl-file-which-feedstock>`__ is ``ExtUtils::MakeMaker``. For typical packaging of packages of this form, you can have a look at `perl-file-which recipe <https://github.com/conda-forge/perl-file-which-feedstock/blob/main/recipe/meta.yaml>`__.

Reviewer: You should also add in information about Module::Build somewhere. Unless you did and I missed it?

Author: Sure, I haven't added it yet. I will read more about it and will add it.

> The most commonly used build system for `perl packages <https://github.com/conda-forge/perl-file-which-feedstock>`__ is ``ExtUtils::MakeMaker``. For typical packaging of packages of this form, you can have a look at `perl-file-which recipe <https://github.com/conda-forge/perl-file-which-feedstock/blob/main/recipe/meta.yaml>`__.

Reviewer: ```suggestion
The most commonly used build system for `Perl packages <https://github.com/conda-forge/perl-file-which-feedstock>`__ is ``ExtUtils::MakeMaker``. For typical packaging of packages of this form, you can have a look at the `perl-file-which recipe <https://github.com/conda-forge/perl-file-which-feedstock/blob/main/recipe/meta.yaml>`__.
```

> The most commonly used build system for `perl packages <https://github.com/conda-forge/perl-file-which-feedstock>`__ is ``ExtUtils::MakeMaker``. For the archetypical packaging of packages of this form, you can have a look at `perl-file-which recipe <https://github.com/conda-forge/perl-file-which-feedstock/blob/main/recipe/meta.yaml>`__.

Reviewer: hyphen needed between commonly and used

Also, do you mean "typical" here, or "archetypal"?

Reviewer: If it's typical, I'd recommend removing the "the". "For typical packaging of packages of this form..."

Author: Thanks for the review Katherine. Actually, I just used the word that was mentioned in the description of the issue, and I didn't know what exactly it meant! ':D
But now that I checked, I think we should just use "typical" as the "archetypal" isn't a very commonly used word so some people might not get it in the first read.

> A few things to note in `perl-file-which recipe <https://github.com/conda-forge/perl-file-which-feedstock/blob/main/recipe/meta.yaml>`__ are:

Reviewer: ```suggestion
A few things to note in the `perl-file-which recipe <https://github.com/conda-forge/perl-file-which-feedstock/blob/main/recipe/meta.yaml>`__ are:
```

>   * The ``name`` is composed of ``perl-`` **+** the lower-cased version of the CPAN name.

Reviewer: ```suggestion
  * The ``name`` is composed of ``perl-`` **+** the lowercase version of the CPAN name.
```

      .. code-block::

          package:
            name: perl-{{ name|lower }}
            version: {{ version }}

>   * The requirements section contain ``make`` for build directive, ``perl`` and ``perl-extutils-makemaker`` for host directive, and ``perl`` for run directive.

Reviewer: ```suggestion
  * The requirements section contain ``make`` for the build directive, ``perl`` and ``perl-extutils-makemaker`` for the host directives, and ``perl`` for the run directive.
```

Reviewer: @beckermr Just checking, am I on the right track with these edits?

Reviewer 2: I am a total perl n00b. Maybe someone from @conda-forge/core can help? @mbargull?

Reviewer 2: We may want to ask the bioconda folks.

Reviewer 2: I bumped bioconda core in the shared gitted channel.

Reviewer 3: Leave it as `host directive` singular, as there is only one `host:` section.

But I would drop “directive” here anyway and mention other dependencies:

> The requirements section contains ``make`` for build, ``perl`` and ``perl-extutils-makemaker`` for host, and ``perl`` for run, in addition to any other packages that the particular Perl module being bullt may require.

      .. code-block::

          requirements:
            build:
              - make
            host:
              - perl
              - perl-extutils-makemaker
            run:
              - perl

>   * The test section contains ``imports`` which lists the CPAN module that can be used in perl.

Reviewer: ```suggestion
  * The test section contains ``imports``, which lists the CPAN module that can be used in perl.
```

>   * The test section contains ``imports``, which lists the CPAN module that can be used in perl.

Reviewer: ```suggestion
  * The test section contains ``imports``, which lists the CPAN module that can be used in Perl.
```

      .. code-block::

          test:
            imports:
              - File::Which

> The build section in the ``meta.yaml`` file should be something like this :

Reviewer: ```suggestion
The build section in the ``meta.yaml`` file should be something like this:
```

.. code-block::

  build:
    number: 0
    noarch: generic
    script:
      - perl Makefile.PL INSTALLDIRS=vendor NO_PERLLOCAL=1 NO_PACKLIST=1
      - make
      - make test
      - make install VERBINST=1

> here the line with ``Makefile.PL`` is important as it guarantees installation into the vendor section, and it also helps in non-interference with the standard files by the ``NO_PERLLOCAL`` and ``NO_PACKLIST`` switches.

Reviewer: ```suggestion
The line with ``Makefile.PL`` is important, as it guarantees installation into the vendor section, and it also helps in non-interference with the standard files by the ``NO_PERLLOCAL`` and ``NO_PACKLIST`` switches.
```

Reviewer: I'm having a hard time understanding this sentence. Does the `Makefile.PL` line guarantee that Perl is installed into the vendor section? Or something else?

And the "non-interference" part of the sentence. What are the standard files and how would those two switches interfere with them?

Reviewer 3: I would rephrase this to discuss the semantic effects desired:

> Notable here are `INSTALLDIRS=vendor`, which selects the appropriate install location for a conda-forge package, and the `NO_PERLLOCAL` and `NO_PACKLIST` switches, which suppress various CPAN tracking files (as described in the `ExtUtils::MakeMaker` documentation) that are not needed when the module is being externally packaged via Conda.

Author: @kathatherine, I also didn't know much about this, this was quite new for me as well. But now what @jmarshall suggested made things a bit more clear. Thanks again John!
 
> `ExtUtils::MakeMaker` documentation

@jmarshall the documentation that you mentioned, is it this one: https://perldoc.perl.org/ExtUtils::MakeMaker? If it is not this can you please share the correct one? It would be nice to add that in the documentation.

Reviewer 3: Yes, that is the correct ExtUtils::MakeMaker documentation.

.. note::

>   ``noarch: generic`` should be used only if the package is a pure perl package.

Reviewer 4: ```suggestion
  ``noarch: generic`` should be used only if the package is a pure Perl package.
```

      .. code-block::

          requirements:
            build:
              - make
            host:
              - perl
              - perl-extutils-makemaker
            run:
              - perl

  * The test section contains ``imports``, which lists the CPAN module that can be used in Perl.

      .. code-block::

          test:
            imports:
              - File::Which

The build section in the ``meta.yaml`` file should be something like this:

.. code-block::

  build:
    number: 0
    noarch: generic
    script:
      - perl Makefile.PL INSTALLDIRS=vendor NO_PERLLOCAL=1 NO_PACKLIST=1
      - make
      - make test
      - make install VERBINST=1

Notable here is ``INSTALLDIRS=vendor``, which selects the appropriate install location for a conda-forge package, and the ``NO_PERLLOCAL`` and ``NO_PACKLIST`` switches, which suppress various CPAN tracking files (as described in the `ExtUtils::MakeMaker documentation <https://perldoc.perl.org/ExtUtils::MakeMaker>`__) that are unnecessary when the module is being externally packaged via conda.

.. note::

  ``noarch: generic`` should be used only if the package is a pure Perl package.

> The other build system used for Perl packages is  ``Module::Build``. ``Module::Build`` is a system for building, testing, and installing Perl modules. 
> It is an alternative to ``ExtUtils::MakeMaker``. It does not require ``make`` - most of the ``Module::Build`` code is pure Perl and written in a very cross-platform way.
> The recipe of some of the Perl packages that uses ``Module::Build`` are :
> 
>   - `perl-math-derivative <https://github.com/bioconda/bioconda-recipes/blob/master/recipes/perl-math-derivative/meta.yaml>`__
>   - `perl-graphics-colornames <https://github.com/bioconda/bioconda-recipes/blob/master/recipes/perl-graphics-colornames/meta.yaml>`__
>   - `mirnature <https://github.com/bioconda/bioconda-recipes/blob/master/recipes/mirnature/meta.yaml>`__
> 
> To know more about ``Module::Build`` and difference between the two build systems read `Comparison <https://metacpan.org/pod/Module::Build#COMPARISON>`__ and `Module::Build <https://metacpan.org/pod/Module::Build>`__.

Reviewer: ```suggestion
The other build system used for Perl packages is  ``Module::Build``, which is a system for building, testing, and installing Perl modules. 
It is an alternative to ``ExtUtils::MakeMaker``. It does not require ``make``; most of the ``Module::Build`` code is pure Perl and written in a very cross-platform way.
The recipes of some of the Perl packages that use ``Module::Build`` are:

  - `perl-math-derivative <https://github.com/bioconda/bioconda-recipes/blob/master/recipes/perl-math-derivative/meta.yaml>`__
  - `perl-graphics-colornames <https://github.com/bioconda/bioconda-recipes/blob/master/recipes/perl-graphics-colornames/meta.yaml>`__
  - `mirnature <https://github.com/bioconda/bioconda-recipes/blob/master/recipes/mirnature/meta.yaml>`__

To know more about ``Module::Build`` and differences between the two build systems, see the `Module::Build <https://metacpan.org/pod/Module::Build>`__ page, paying special attention to the `Comparison <https://metacpan.org/pod/Module::Build#COMPARISON>`__ section.
```

      - make test
      - make install VERBINST=1

Notable here is ``INSTALLDIRS=vendor``, which selects the appropriate install location for a conda-forge package, and the ``NO_PERLLOCAL`` and ``NO_PACKLIST`` switches, which suppress various CPAN tracking files (as described in the `ExtUtils::MakeMaker documentation <https://perldoc.perl.org/ExtUtils::MakeMaker>`__) that are unnecessary when the module is being externally packaged via conda.

.. note::

  ``noarch: generic`` should be used only if the package is a pure Perl package.

The other build system used for Perl packages is  ``Module::Build``. ``Module::Build`` is a system for building, testing, and installing Perl modules. 

> It is an alternative to ``ExtUtils::MakeMaker``. It does not require a ``make`` on your system - most of the ``Module::Build`` code is pure-perl and written in a very cross-platform way.

Reviewer 5: ```suggestion
It is an alternative to ``ExtUtils::MakeMaker``. It does not require ``make`` - most of the ``Module::Build`` code is pure Perl and written in a very cross-platform way.
```

I would add some examples here (maybe from `bioconda`). A quick search on the bioconda org shows [several useful results](https://cs.github.com/?scopeName=All+repos&scope=&q=org%3Abioconda+perl-module-build+path%3Ameta.yaml). For example:

* https://github.com/bioconda/bioconda-recipes/blob/master/recipes/perl-math-derivative/meta.yaml

> To know more about ``Module::Build`` and difference between the two build systems read `Comparision <https://metacpan.org/pod/Module::Build#COMPARISON>`__ and `Module::Build <https://metacpan.org/pod/Module::Build>`__.

Reviewer 5: ```suggestion
To know more about ``Module::Build`` and difference between the two build systems read `Comparison <https://metacpan.org/pod/Module::Build#COMPARISON>`__ and `Module::Build <https://metacpan.org/pod/Module::Build>`__.
```
